test/dynamo/test_misc.py

```diff
@@ -7848,6 +7848,42 @@ def pow(x):
             msg="Encountered an unexpected fallback to 'aten pow' in dynamo compiled code",
         )
 
+    def test_symint_key(self):
+        class Mod(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x, y):
+                emb_dim_sizes: t.Dict[int, bool] = {}
+                for shape in [x.shape, y.shape]:
+                    emb_dim_sizes[shape[0]] = True
+                if len(emb_dim_sizes) != 1:
+                    raise RuntimeError(
+                        f"Found different dim sizes in all inputs: {emb_dim_sizes.keys()=}"
+                    )
+                return x + y
+
+        mod = Mod()
+        x = torch.rand(4, 5)
+        y = torch.rand(4, 5)
+        eager_out = mod(x, y)
+
+        cnts = torch._dynamo.testing.CompileCounter()
+        mod = torch._dynamo.optimize(cnts, nopython=True)(mod)
```

Comment:
Discussed offline - your program specializes on the value, it cannot both specialize AND fully guarantee your desired range. Something has to give, unless you want to go ahead and write a series of higher order ops to work around this.

Reply:
Can you help explain this?

> it cannot both specialize AND fully guarantee your desired range

On the high level, I want to tell dynamo that "my input tensors 0-th shape is subject to change within the range of [1,16]". How am I going to convey that in terms of `export__RC__` API? Thanks! 


```
#!/usr/bin/env python3

import typing as t
import torch

class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, x, y):
        emb_dim_sizes: t.Dict[int, bool] = {}
        for shape in [x.shape, y.shape]:
            emb_dim_sizes[shape[0]] = True
        if len(emb_dim_sizes) != 1:
            raise RuntimeError(f"Found different dim sizes in all inputs")
        return x + y


mod = Mod()
x = torch.rand(4,5)
y = torch.rand(4,5)
mod(x, y)  # OK


def main():

    compiled_mod = torch._export.export__RC__(
        mod, (x, y),
        dynamic_shapes={
            'x': [torch.export.Dim(name='x', min=1, max=16), None],
            'y': [torch.export.Dim(name='y', min=1, max=16), None],
        },
    )

    # test run with changed batch size: 4 -> 6
    # this fails without setting "dynamic_shapes" above
    compiled_mod(
        torch.rand(6,5),
        torch.rand(6,5),
    )

main()
```

Reply:
Following my previous comment:

> I think the key is if there's a way we can tell dynamo not to over-specialize the sample input shape-0 to a constant 4 (as mentioned in the error msg).
> 
> > Not all values of x = L['x'].size()[0] in the specified range x <= 16 are valid because x was inferred to be a constant (4).


I wonder if I can construct a sample input with symbolic shape? like

```
batch_dim = torch.SymInt(sympy.Symbol('batch_dim'))
sample_input_x = torch.rand(batch_dim, 3)
```

Reply:
Oh, @kflu, this is the logic in torchrec, isn't it?

The long term correct solution in that case is to unconditionally use the variable batch size logic, skipping the "emb_dim_sizes" hash match. You can see me hacking this up in e.g. at https://github.com/pytorch/torchrec/compare/main...ezyang:torchrec:dlrm-blaze-checkpoint6#diff-9b80c65f66abaf80b327456264ca7b031c6fa424946c489098b45aecf9b82c97R1858

Reply:
We spoke about this a bit - I think there's a HOP + meta solution here as well, let me ping you.

Reply:
Hi @eddieluo01 this is an internal model that I tried to enable AOT inductor. If I follow your thoughts, that means I should remove those runtime validations inside my model. 

one question - on what level dynamo can provide these type of shape compatibility checks for the users, so I can safely remove them from the model code?

And also, I tried to remove the shape checks in my code, however, I still get an error regarding the dynamic_shape config:

```
#!/usr/bin/env python3

import typing as t
import torch


class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, a, b, c):
        n_vectors = f([a,b,c])
        return torch.cat((a,b,c), dim=1).reshape(-1, n_vectors)

def f(embeddings: t.List[torch.Tensor]):
    all_shapes = [_.shape for _ in embeddings]
    input_shape = all_shapes[0]
    if len(input_shape) != 2 and len(input_shape) != 3:
        raise RuntimeError("2 or 3")

    if len(input_shape) == 2:
        return len(embeddings)
    else:
        n_vector = 0
        for shape in all_shapes:
            n_vector += shape[1]
        return n_vector



mod = Mod()
inp = (
    torch.rand(4,5),
    torch.rand(4,5),
    torch.rand(4,5),
)
mod(*inp)



compiled_mod = torch._export.export__RC__(
    mod, inp,
    dynamic_shapes={
        'a': [torch.export.Dim(name='a', min=1, max=16), None],
        'b': [torch.export.Dim(name='b', min=1, max=16), None],
        'c': [torch.export.Dim(name='c', min=1, max=16), None],
    },
)
# Error: 
UserError: Constraints violated (b, c)! For more information, run with TORCH_LOGS=dynamic.
  - The values of b = L['b'].size()[0] and a = L['a'].size()[0] must always be equal.
  - The values of c = L['c'].size()[0] and a = L['a'].size()[0] must always be equal.

Suggested fixes:
  c = a
  b = a
```

Reply:
The error message here seems accurate. Why are you allocate separate dims for a b and c? Your program fails when I pass it

```
mod = Mod()
inp = ( 
    torch.rand(4,5),
    torch.rand(4,5),
    torch.rand(4,5),
)
mod(*inp)
```

Reply:
@avikchaudhuri do you think there is a way to make the error message better?